### DISCUSSION TITLE
Multi-line descriptions do not import

### DISCUSSION THREAD

**Discussion by calebdw** [spatie/icalendar-generator#84] [Bugs]
When adding a multi-line description to an Event, only the first line appears in Outlook. Line breaks also appear in odd places.
I saved a random event from Outlook as an .ics, and the entire description block was tab-indented as shown below (which this library does not do):

```
DESCRIPTION:\n\n-----Original Appointment-----\n
	Please join my meeting from your computer\, tablet or sma
	rtphone.\n\nGet the app now and be ready when your first meeting starts
DTEND;TZID="Central Standard Time":20220308T120000
```

**Comment by rubenvanassche** [comment]
@CalebDW Could you give me an example of such an event in PHP?

We have chipping of lines in place because it is required in the spec (https://github.com/spatie/icalendar-generator/blob/a65dc5cb8c1c0ecfd68adf9f52be164141e41c88/tests/Builders/ComponentBuilderTest.php#L79)

**Reply by calebdw** [reply]
@rubenvanassche,
 I took your really long test string, created a test meeting in Outlook with it, and then exported the meeting to a .ics file. This is what the top of the file looked like:
```
BEGIN:VCALENDAR
PRODID:-//Microsoft Corporation//Outlook 16.0 MIMEDIR//EN
VERSION:2.0
METHOD:REQUEST
X-MS-OLK-FORCEINSPECTOROPEN:TRUE
BEGIN:VTIMEZONE
TZID:Central Standard Time
BEGIN:STANDARD
DTSTART:16011104T020000
RRULE:FREQ=YEARLY;BYDAY=1SU;BYMONTH=11
TZOFFSETFROM:-0500
TZOFFSETTO:-0600
END:STANDARD
BEGIN:DAYLIGHT
DTSTART:16010311T020000
RRULE:FREQ=YEARLY;BYDAY=2SU;BYMONTH=3
TZOFFSETFROM:-0600
TZOFFSETTO:-0500
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VEVENT
ATTENDEE;CN=me@mycompany.com;RSVP=TRUE:mailto:me@mycompany.com
CLASS:PUBLIC
CREATED:20220328T195812Z
DESCRIPTION:This is a really long text. Possibly you will never write a tex
	t like this in a property. But hey we support the RFC so let us chip it! Y
	ou can maybe write some HTML in here that will make it longer than usual.\
	n\n
DTEND;TZID="Central Standard Time":20220329T085000
DTSTAMP:20220328T195745Z
DTSTART;TZID="Central Standard Time":20220329T080000
LAST-MODIFIED:20220328T195812Z
LOCATION:Online
```

If I remove the tabs in the description so that it becomes:
```
DESCRIPTION:This is a really long text. Possibly you will never write a tex
t like this in a property. But hey we support the RFC so let us chip it! Y
ou can maybe write some HTML in here that will make it longer than usual.\
n\n
DTEND;TZID="Central Standard Time":20220329T085000
```
then all that is imported into Outlook is: `This is a really long text. Possibly you will never write a tex`.

It seems to me like your line-chipping needs to prepend a tab character to the start of every chipped line (or rather use '\n\t' instead of '\r'...or something like that) so that Outlook will recognize it as a block of continuous text. Have you successfully imported a calendar event with really long lines into Outlook?

**Reply by rubenvanassche** [reply]
Our chipping mechanism actually doesn't use tabs: https://github.com/spatie/icalendar-generator/blob/main/src/Builders/ComponentBuilder.php#L87. I do not have Outlook so I'm unable to test what's actually going wrong.

**Reply by calebdw** [reply]
Perhaps it should use tabs instead of spaces (change `' '` to `"\t"`)?

**Reply by rubenvanassche** [reply]
The RFC explicitly says to use spaces: https://datatracker.ietf.org/doc/html/rfc5545#section-3.3.11

I've checked some implementations of other packages and they alm use the same chipping mechanism:

- https://github.com/ZContent/icalendar/blob/master/includes/ical.php#L435-L458
- https://github.com/markuspoerschke/iCal/blob/2.x/src/Presentation/ContentLine.php

I'm not sure but since you're using newlines within the text this could maybe break the implementation somehow

**Reply by calebdw** [reply]
The RFC does allow both a space OR a tab: https://datatracker.ietf.org/doc/html/rfc5545#section-3.1

>   Lines of text SHOULD NOT be longer than 75 octets, excluding the line
   break.  Long content lines SHOULD be split into a multiple line
   representations using a line "folding" technique.  That is, a long
   line can be split between any two characters by inserting a CRLF
   immediately followed by a single linear white-space character (i.e.,
   SPACE or HTAB).  Any sequence of CRLF followed immediately by a
   single linear white-space character is ignored (i.e., removed) when
   processing the content type.

But it looks like the CRLF (\r\n) must be appended to the end of the line which I don't see in your source, but these guys are doing it here:
https://github.com/ZContent/icalendar/blob/master/includes/ical.php#L435-L458

Also note that the first example from Outlook I posted has newlines and it works so their existence is not the issue, but do note that the folks above convert any CRLF/CR values to LF before appending the CRLF char:
```php
	function printDataLine($d, $p)
	{
                // ...
		$line = str_replace(array("<br>","<BR>","<br/>","<BR/"),"\\n",$line);
		$line = str_replace(array("\r\n","\n\r","\n","\r"),'\n',$line);
                // ...
		while (strlen($line) > 0) {
			$linewidth = ($linecount == 0? 75 : 74);
			$linesize = (strlen($line) > $linewidth? $linewidth: strlen($line));
			if($linecount > 0)
				$txtstr .= " ";
			$txtstr .= substr($line,0,$linesize) . "\r\n";
                // ...
	}
```

**Reply by rubenvanassche** [reply]
We add these characters over here: https://github.com/spatie/icalendar-generator/blob/main/src/Builders/ComponentBuilder.php#L29 So that could not be the issue. At the moment we only convert characters defined by the RFC, maybe that set should be extended: https://github.com/spatie/icalendar-generator/blob/main/src/Properties/TextProperty.php#L42-L48

**Reply by maxresnikoff** [reply]
I am also having this issue. `"spatie/icalendar-generator": "^2.6"`

When doing multi-line descriptions, the encoded output seems to be quite messy.

It seems like when the encoding happens, it adds `\n` and then a `CR`
As mentioned by @calebdw, it requires a CRLF at the end of the line followed by a single white space.

I made the following changes to the way I set the description, which might hopefully help come up with a fix for this issue: 

- adding in the `\n` manually into the description string instead of using physical new lines in the script.
- setting a custom TextProperty for the description
- adding `withoutEncoding()` to this new TextProperty

It then works without issues!

Instead of this:

```
$event->description($description)
```

Replace with:
```
$descriptionProp = TextProperty::create('DESCRIPTION', $description)->withoutEscaping();
$event->appendProperty($descriptionProp);
```

Using the default `->description()` renders this in the ics file:
![image](https://github.com/spatie/icalendar-generator/assets/48560713/f871f9f0-954b-49df-8549-8bcd6dc627fe)


Using the `TextProperty` solution, it renders correctly:
![image](https://github.com/spatie/icalendar-generator/assets/48560713/a2d1c4eb-b838-473d-856f-cf81c5c322ff)

**Comment by seem2810** [comment]
Hello,
I've also a problem with the new lines on iphones.
The events wouldnt be importet.
What do I wrong?

Long descriptions, eG:
```
BEGIN:VEVENT
UID:iCal17
DTSTAMP:20220602T210855Z
SUMMARY:5x10m in xxxxx -
DESCRIPTION:1. Wann wird es benötigt?  - 2. Juli\n2. Wo wird es benötigt
  Straße/Ort? xxxx\n3. Wann kann Auf-/Abgebaut werden? Gerne am Tag
  vorher und nachher. Wir sind flexibel \n4. Welche Größe benötigen Sie
  für Ihren Anlass? Keine Ahnung. Wir möchten mit 50 Personen feiern. Wel
 che Größe wird benötigt? \n5. Zufahrt mit den PKW-Anhänger/LKW Mögli
 ch? Nicht ganz bis zum Aufbauplatz. Ca. 20 Meter muss manuell transportier
 t werden. . \n6. Falls das Zelt ohne Boden gewählt werden soll\, kann di
 eses mittels Erdnägel fixiert werden? - ja. Sind unter der Erde Leitungen
 /Rohre bekannt\, die im Weg sein könnten?  Nein\n7. Falls das Zelt jedoc
 h mit Boden ausgewählt werden sollte\, muss dieser Ausgerichtet werden? W
 enn ja\, in welchem Ausmaß?  Ja\, ca20 cm? Es ist für mich schwer abzusc
 hätzen \n8. Können Sie uns Fotos/Video des Aufbauorts-/Zufahrt(Zugang) 
 zusenden? Ja\, später \n9. Wird noch weiteres Zubehör wie Musikanlage\,
  Lichteffekte oder ähnliches benötigt? Evtl.\, es hängt vom Angebot ab
 \n10. Ihre speziellen Wünsche\, die den Auftrag beeinflussen gerne unter 
 diesem Punkt erwähnen :-).  Wir können gerne Hilfe leisten\, damit es f
 ür uns günstiger sein wird \n\n xxxxx21\, xxxxxAm Ende 
 der Sackgasse rechts\, das Haus mit einem Metall-Tor\n\n0151xxxxxxx\n
 \nHallo Herr XXX\, \nwenn Sie hierfür eine Rechnung machen\, dann bi
 tte an \nxxxx\nMxxxxr 9\nxxxxx. \n\nSie err
 eichen mich unter xxx@exxxx\n\nDie Anlieferadres
 se\, wie besprochen\, XXXX
DTSTART;TZID=Europe/Berlin:20220630T180000
DTEND;TZID=Europe/Berlin:20220704T180000
END:VEVENT
BEGIN:VEVENT
```

or
```
BEGIN:VEVENT
UID:iCal10
DTSTAMP:20220424T141259Z
SUMMARY:5x10m (Boden) + PA-Anlage (Mixer+Pult) + Partybar + LED Zelt Beleuc
 htung
DESCRIPTION:Preis 950-1000\,-€\n\n5x10m inkl. Boden - PA-Anlage inkl. M
 ischpult und Mixer - Partybar und LED Zeltbeleuchtung. \n\nAufbau : 18.0
 8.2022 ab 18:00\nAbbau: Sonntag 21.08 16:00 oder 22.08 ab 18:00 (wenn es 
 Sonntag durch uns nicht klappt) \n\nAufbauort: XXXXX5\, XXXXX
 XXXXX\n\nAnsprechpartner:\n\nXXXX XXXXXp\nÜXXXX 5\nXXXXX
  XXX\n0171/XXXXX\n02XXXXX\n\nMarder Netz im Boden\, welches
  kaputt gehen wird durch das befestigen mit Erdnägel (Mieter weiss besche
 id). Mehraufwand durch Transport ums Haus.
DTSTART;TZID=Europe/Berlin:20220818T180000
DTEND;TZID=Europe/Berlin:20220822T180000
END:VEVENT
```

After Adding a substr it works..
```
$event->description = str_replace(PHP_EOL, ' ', $event->description);
$event->description = str_replace('€', 'EUR', $event->description);
$event->description = substr($event->description, 0, 50);
```
